data/virt_autotest/fetch_logs_from_guest.sh

```diff
@@ -0,0 +1,260 @@
+	local logs_folder=$3
+
+        setup_libguestfs_env
+	virt-filesystems -d ${guest_domain} &> /dev/null
+	if [[ $? -ne 0 ]];then
+	   echo -e "Running ${guest_domain} can not be accessed by libguestfs currently. Shut it down now."
+	   virsh destroy ${guest_domain}
+	fi
+	local guest_device=`find_disk_hosts_filesystem ${guest_domain} ${logs_fetched}`
+        echo -e "${guest_domain} ${guest_device} contains ${logs_fetched}."
+        local guest_transformed=${guest_domain//./_}
+        mkdir -p ${logs_folder}/${guest_transformed}
+        echo -e "guestfish -r -d ${guest_domain} -m ${guest_device} copy-out ${logs_fetched} ${logs_folder}/${guest_transformed}"
+        guestfish -r -d ${guest_domain} -m ${guest_device} copy-out ${logs_fetched} ${logs_folder}/${guest_transformed}
+        if [[ $? -eq 0 ]];then
+           echo -e "Copied out ${logs_fetched} from ${guest_domain} successfully."
+           guestfish -r -d ${guest_domain} -m ${guest_device} rm-rf ${logs_fetched}
+           return 0
+        else
+           echo -e "Try again to mount ${guest_device}:/:subvol=@ with guestfish."
+           echo -e "guestfish -r -d ${guest_domain} -m ${guest_device}:/:subvol=@ copy-out ${logs_fetched} ${logs_folder}/${guest_transformed}"
+           guestfish -r -d ${guest_domain} -m ${guest_device}:/:subvol=@ copy-out ${logs_fetched} ${logs_folder}/${guest_transformed}
+           if [[ $? -eq 0 ]];then
+              echo -e "Copied out ${logs_fetched} from ${guest_domain} successfully."
+              guestfish -r -d ${guest_domain} -m ${guest_device}:/:subvol=@ rm-rf ${logs_fetched}
+              return 0
+           else
+              echo -e "Failed to copy out ${logs_fetched} from ${guest_domain}."
+              return 1
+           fi
+	fi
+}
+
+#Fetch logs from virtual machine to local host via ssh firstly. Will resort to libguestfs tools if ssh connection is broken
+function fetch_logs_from_guest() {
+	local guest_domain=$1
+	local guest_ipaddr=$2
+	local logs_folder=$3
+	shift
+	shift
+	shift
+	local extra_logs=($@)
+
+	local ret1=0
+	fetch_logs_from_guest_via_ssh ${guest_domain} ${guest_ipaddr} ${logs_folder}
+	if [[ $? -ne 0 ]];then
+	   echo -e "Try to use libguestfs tools to fetch ${logs_folder} from ${guest_domain}."
+           fetch_logs_from_guest_via_libguestfS ${guest_domain} ${logs_folder} ${logs_folder}
+	   ret1=$?
+	fi
+
+	local ret2=0
+	if [[ ${extra_logs[@]} != "" ]];then
+           echo -e "Try to use libguestfs tools to fetch ${extra_logs[@]} from ${guest_domain}."
+           for eachlog in ${extra_logs[@]};do
+	       fetch_logs_from_guest_via_libguestfS ${guest_domain} ${eachlog} ${logs_folder}
+	       ret2=$(( ${ret2} | $? ))
+	   done           
+	   if [[ ${ret2} -eq 0 ]];then
+	      echo -e "Copied out ${extra_logs[@]} from ${guest_domain} successfully."
+	   else
+	      echo -e "Failed to copy out ${extra_logs[@]} from ${guest_domain}."
+	   fi
+	fi
+        
+	virsh start ${guest_domain}
```

Comment:
Is this start really needed? If it helps for no purpose, suggest to remove it since it takes resource for nothing.

Reply:
Somethings virtual machine needs to be shutoff before performing files copy-out. So ensure it is active at the last.

Reply:
I know to copy-out , the guest maybe shut down. You want to recover it. But this action happens in post_fail_hook which is at the end of the test. After it, the SUT(host) will be shutoff, so does all vms. Then why bother to start vm again? This is my point. So I ask you what purpose(other things intended to be performed on guest) is the start of vm here. If no, then keeping it off is just fine -- save os resources in multiple vms test scenario.

Reply:
I just want them back to their original state. I can keep them off, pretty easy.

Reply:
One more thing is to maintain test continuity. So not all test modules will reboot guest at the start. So let's keep this "virsh start" here. It is useful